docs: add local unit test and linting instructions to CONTRIBUTING.md (#3035)
ansible-collections/amazon.aws: CONTRIBUTING.md

If you're struggling with running integration tests locally, don't worry.
After creating a pull request the GitHub Actions will automatically test for you.

## Running Tests Locally

### Unit Tests

Unit tests are run via [tox](https://tox.wiki/). The collection uses fully-qualified
`ansible_collections.amazon.aws.*` imports, which require the Ansible collection loader
to resolve the namespace correctly. `tox` handles this automatically by syncing the
collection into a managed environment before running `pytest`.

First, install `tox` into your virtual environment:

```bash
pip install tox
```

To run all unit tests:

```bash

> tox -e ansible2.20-py312-without_constraints

Reviewer: alternatively

`tox -m unit-oldest` or `tox -m unit-newest`

> ```

Reviewer 2: I feel like this is somewhat misleading. I ran into this myself recently while looking at a pull request in this collection. There are multiple combinations of `ansible-core` and Python versions in the `envlist` and CI runs the lot.

Reading "all unit tests" might cause contributors to think they've fully covered their changes and not just one combo. So I'd rephrase to say something like:

"To run a specific environment that matches your Python version:" or something like that.

For all unit tests, I believe you would need `tox -m unit`.

To run a specific test file or filter by test name, use the `-k` flag rather than
passing a file path directly (file paths resolve relative to the repo root and break
the collection namespace resolution inside tox):

```bash
tox -e ansible2.20-py312-without_constraints -- -k test_aws_sqs_queue -v
```

See `tox.ini` for the full list of available environments (ansible-core versions and

> Python version combinations).

Reviewer 2: What about recommending `tox list` to discover available environments? That might be more straightforward for folks than digging through `tox.ini`.

> **Note:** Do not use `pytest` or `ansible-test units` directly — they will fail with
> `ModuleNotFoundError: No module named 'ansible_collections'` because the collection
> namespace is not on `sys.path` without the tox setup.

> 

Reviewer 2: I feel like `tox -e diff-cover` is worth a mention in here. https://github.com/ansible-collections/amazon.aws/blob/14294b0f70cb0a8c05cf8f71e4b407125acfe910/tox.ini#L89

### Linting

> Two linters are enforced in CI. Run them against changed files before submitting a PR:

Reviewer 2: I believe that there are other linters that run in CI. Looking at: https://github.com/ansible-collections/amazon.aws/blob/14294b0f70cb0a8c05cf8f71e4b407125acfe910/.github/workflows/linters.yml#L11

That should run all the linters with `labels = lint` in the tox config such as https://github.com/ansible-collections/amazon.aws/blob/14294b0f70cb0a8c05cf8f71e4b407125acfe910/tox.ini#L129

So this section could be reduced down to `tox -m lint` instead of individual commands.

**Black** (code formatting):

```bash
tox -e black-lint -- <path/to/file>
```

**Flake8** (style and error checking):

```bash
tox -e flake8-lint -- <path/to/file>
```

> To auto-fix formatting issues (rather than just checking), use the `black` environment:

Reviewer 2: ```suggestion
To automatically fix any formatting issues, use the following command:
```

```bash

> tox -e black -- <path/to/file>

Reviewer: Even simpler:

```suggestion
tox -m format
```
